### PR TITLE
Create general Cache Service

### PR DESCRIPTION
**Pull request by MarcelTorne**
Refactoring code from PoiCachedService into a general cache service that can be reused by different activities and classes. I am also creating an interface `ResponseFetcher` which is the service used by the cache service to get the response in case it is not found in the local cache.

### CODE REVIEW

File: app/src/main/java/com/github/epfl/meili/cache/CacheService.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    protected open fun isCacheValid(cachedTimestamp: Long): Boolean {
+        return verifyValidity(cachedTimestamp)
+    }
+
+    /**
+     * Data is considered valid if it was retrieved less than 1 hour ago and the new request comes
+     * from less than 1km far away from the last data
+     *
+     * @return whether the data in the object is valid or not
+     */
```
Comment (aman3014) [inline app/src/main/java/com/github/epfl/meili/cache/CacheService.kt:139]: Leftover comment

File: app/src/main/java/com/github/epfl/meili/cache/CacheService.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    private fun retrieveTimeOfResponse(): Long {
+        return mPrefs.getLong(TIMESTAMP_KEY, 0L)
+    }
+
+    /**
+     * Data is considered valid if it was retrieved less than 1 hour ago and the new request comes
+     * from less than 1km far away from the last data
+     *
+     * @return whether the data saved in shared preferences is valid or not
+     */
```
Comment (aman3014) [inline app/src/main/java/com/github/epfl/meili/cache/CacheService.kt:129]: Leftover comment

**Review by aman3014 — APPROVED** [review]
Looks great, thanks!

File: app/src/main/java/com/github/epfl/meili/cache/CacheService.kt (hunk 2 above)
Comment (MarcelTorne) [reply app/src/main/java/com/github/epfl/meili/cache/CacheService.kt:128]: Changed! Thanks for pointing it out!

**Review by Bwananah — APPROVED** [review]
Very clean, good job !

**Pull request merged**